### ISSUE TITLE
ClassFanOutComplexity: false positive on var

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#6346]
Taken from https://github.com/checkstyle/checkstyle/issues/6316#issuecomment-451522961

````
$ cat TestClass.java
public class TestClass {
    void method() {
var x = 1;
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="ClassFanOutComplexity">
  <property name="max" value="0"/>
</module>
    </module>
</module>

$ java -jar checkstyle-8.15-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:1:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
````

`var` should not be affecting ClassFanOutComplexity by default.
With `var` changed to `Object`, no violation is produced.

https://checkstyle.org/config_metrics.html#ClassFanOutComplexity should ignore var,int type should be counted ... as it is not a class..

**Comment by strkkk** [comment]
@rnveach 
This check should ignore `var` completely?
if not, issue requires clarification - if right side is, for example, a method invocation
```
var myVar = godKnowsWhatThisMethodReturns();
```
I guess we cannot infer type here.

The only case when we know type is a String/int/char/etc. literal on right side.

**Comment by rnveach** [comment]
@romani Please review.

> This check should ignore var completely?

My initial response is yes.

Based on the JLS,
https://docs.oracle.com/javase/specs/jls/se10/html/jls-14.html#jls-14.4.1
>  If LocalVariableType is var, then let T be the type of the initializer expression when treated as if it did not appear in an assignment context, and were thus a standalone expression (§15.2). The type of the local variable is the upward projection of T with respect to all synthetic type variables mentioned by T (§4.10.5). 

I would say it should behave the same as `T` as if it was a template for a method/class.
> `<T> void method(T param) {}`

But I checked this out with the check, and it flags template classes too.
@romani Do you agree that template classes shouldn't be flagged too?

````
$ cat TestClass.java
public class TestClass {
    <T> void method(T param) {
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="ClassFanOutComplexity">
  <property name="max" value="0"/>
</module>
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:1:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
````

**Comment by strkkk** [comment]
@rnveach What about ClassDataAbstractionCouplingCheck? It has same superclass as ClassFanOutComplexity and probably `var` exclusion should be considered for it too.

**Comment by romani** [comment]
>I guess we cannot infer type here.

Checkstyle can not infer types. Checkstyle works with string values. "var" for us is the same as Object.

>Do you agree that template classes shouldn't be flagged too?

It is ok, to be counted, as it some type and it is distinct. So  amount of types is  amount of tepmplate names.

Var is different for us, as amount of vars, it not amount of distinct types.
So it is better to ignore it completely.

**Comment by strkkk** [comment]
@romani ClassFanOutComplexity and ClassDataAbstractionCoupling have same superclass and same list of ignored classes.
var can be added to this list and it looks like it will solve this issue, but there is side effect on ClassDataAbstractionCouplingCheck, it will ignore cases like:
```
Object a = new var();
``` 

where var is a class name. It is not forbidden in java9 and lower. Looks weird and breaks java naming convention, but still possible.
In my opinion, there are two options - add note to ClassDataAbstractionCouplingCheck or change existing design to add var only to ClassFanOutComplexity default excluded classes.

**Comment by romani** [comment]
if we add `var` to excluded classes, it will force it to appear in documentation. `var` is not a class for users. It is internal checkstyle's nuance that we treat it same as type. Ideally we should update grammar to keep it part of tokens.

both such Checks should ignore `var` (in definition/left side) to my mind.

**Comment by romani** [comment]
Fix is Merged
